File: DIP/C#/não aplicado/Pedido.cs
public enum FormasPagamento
{
	CARTAOCREDITO
}

public class Pedido
{
	readonly NotificadorService _notificador;
	readonly ProcessadorPagamentoService _processador;
	readonly ReservadorEstoqueService _reservador;
	
	public Pedido()
	{
		_notificador = new NotificadorService();
		_processador = new ProcessadorPagamentoService();
		_reservador = new ReservadorEstoqueService();
	}
	
	public void Comprar(Carrinho carrinho, DetalhesPagamento detalhesPagamento, bool deveNotificarCliente)
	{
		if(detalhesPagamento.FormaPagamento == FormasPagamento.CARTAOCREDITO)
			_processador.Processar(detalhesDoPagamento, carrinho);
			
		_reservador.ReservarEstoque(carrinho);
		
		if(deveNotificarCliente) 
			_notificador.NotificarCliente(carrinho);
	}
}

File: DIP/Ruby/não aplicado/pedido.rb
require 'ostruct'
require 'net/smtp'
require 'notificadorService'
require 'processadorPagamentoService'
require 'reservadorEstoqueService'

class FormasPagamento
    CARTAOCREDITO = 1   
end

class Pedido
    
    def initialize
        @reservadorEstoque = ReservadorEstoqueService.new
        @notificador = NotificadorService.new
        @processadorPagamento = ProcessadorPagamentoService.new
    end

    def comprar(carrinho, detalhesDoPagamento, deveNotificaCliente)
        @processadorPagamento.processar(detalhesDoPagamento, carrinho.valorTotal) if detalhesDoPagamento.formaPagamento == FormasPagamento::CARTAOCREDITO
        
        @reservadorEstoque.reservar(carrinho)

        @notificador.notificarCliente(carrinho) if deveNotificaCliente
    end 
end